stellargraph/layer/rgcn.py

```diff
@@ -322,8 +316,14 @@ def call(self, inputs):
         # Calculate the layer operation of RGCN
         output = K.dot(features, self.self_kernel)
 
+        if self.num_bases > 0:
+            self.relational_kernels = [
```

Comment:
It's a little weird to me that this is mutating the `self` object, and saving these computed values outside this `call` method. In particular, if doing this computation has to happen inside the `call` method, it might not be valid to use the results after the `call` method, and storing them in `self` is risking hitting that problem.

As such, maybe:

```python
if self.num_bases > 0:
    relational_kernels = [ ... ]
else:
    relational_kernels = self.relational_kernels

for i in range(self.num_relationships):
    h_graph = K.dot(...)
    output += K.dot(..., relational_kernels[I])
```

But... possibly my suspicion is wrong and it's totally fine and even useful to hold the result for longer. If that's the case, then maybe we could reduce how much work we do here (assuming `call` is called more than once) by not recomputing the kernels:

```python
if self.relational_kernels is None:
    assert self.num_bases > 0, "delayed initialisation only works for basis-matrix formulation"
    self.relational_kernels = [ ... ]
```

Reply:
I tried this but it yielded unconnected gradient errors (weirdly different to the prior warnings). `self.relational_kernels` is the product of trainable variables so it has to be recalculated every training step. 

My suspicion is that in non-eager mode this can be done implicitly by baking this calculation in the computational graph but in eager mode this recalculation has to done be explicitly. Anyway, I'll try and neaten this sectino up a bit

Reply:
> I tried this

My comment had two suggestions, so I'm not sure what "this" means. From context it sounds like the second suggestion? In which case, it does sound like using the results of these computations outside `call` is incorrect and so I would recommend trying something along the lines of the first suggestion.

Reply:
ah sorry, I was referring to your second suggestion. I've implemented your first suggestion which really tidies up the call function.